### PR TITLE
[mono] Don't use recursion in mono_ssa_rename_vars

### PR DESCRIPTION
**Pull request by radekdoulik**
Implements https://github.com/dotnet/runtime/issues/60266

Avoid deep recursion in `mono_ssa_rename_vars`, change the way we traverse
dominated bb's. Instead of using recursion, use stack-like array to
store information about stack history and the traversal.

The performance remains the same (or is slightly better) than before.
Times for the compilation of repro from https://github.com/dotnet/runtime/issues/57141
(JIT time minimum from 5 runs):

Before:

    LLVM output file: './Uno.UI.FluentTheme.dll.bc.tmp'.
    JIT time: 4810 ms, Generation time: 2040 ms, Assembly+Link time: 2 ms.

After:

    LLVM output file: './Uno.UI.FluentTheme.dll.bc.tmp'.
    JIT time: 4781 ms, Generation time: 2017 ms, Assembly+Link time: 2 ms.

### CODE REVIEW

**Review by lewing — APPROVED** [review]
This looks ok to me, it would be nice if someone on the runtime side took look.

**Comment by marek-safar** [comment]
@vargaz @imhameed does this look ok?

File: src/mono/mono/mini/ssa.c [hunk 1]
```diff
@@ -159,24 +159,47 @@ typedef struct {
...
+			}
+			else
+				/* The phi nodes are at the beginning of the bblock */
+				break;
+		}
+	}
+}
+
+static int
+create_new_vars(MonoCompile *cfg, int max_vars, MonoBasicBlock *bb, gboolean *originals_used, MonoInst **stack, guint32 *lvreg_stack, gboolean *lvreg_defined, RenameInfo **stack_history, int *stack_history_size)
```
Comment (imhameed) [inline src/mono/mono/mini/ssa.c:198]:
```suggestion
create_new_vars (MonoCompile *cfg, int max_vars, MonoBasicBlock *bb, gboolean *originals_used, MonoInst **stack, guint32 *lvreg_stack, gboolean *lvreg_defined, RenameInfo **stack_history, int *stack_history_size)
```

File: src/mono/mono/mini/ssa.c [hunk 2]
```diff
@@ -284,49 +307,103 @@ mono_ssa_rename_vars (MonoCompile *cfg, int max_vars, MonoBasicBlock *bb, gboole
...
 
+	return stack_history_len;
+}
+
+static void
+restore_stack (MonoInst **stack, RenameInfo *stack_history, int stack_history_len)
+{
+	int i;
+
+	for (i = stack_history_len - 1; i >= 0; i--) {
```
Comment (imhameed) [inline src/mono/mono/mini/ssa.c:320]:
```suggestion
	int i = stack_history_len;
	while (i-- > 0) {
```

File: src/mono/mono/mini/ssa.c [hunk 3]
```diff
@@ -284,49 +307,103 @@ mono_ssa_rename_vars (MonoCompile *cfg, int max_vars, MonoBasicBlock *bb, gboole
...
+	stack_history = g_new (RenameInfo, stack_history_size);
+	originals = g_new0 (gboolean, cfg->num_varinfo);
+	rename_stack_size = 16;
+	rename_stack = g_new (RenameStackInfo, rename_stack_size);
+
+	do {
+		if (G_UNLIKELY (cfg->verbose_level >= 4))
+			printf ("\nRENAME VARS BLOCK %d:\n", bb->block_num);
+
+		int stack_history_len = create_new_vars (cfg, max_vars, bb, originals,  stack, lvreg_stack, lvreg_defined, &stack_history, &stack_history_size);
```
Comment (imhameed) [inline src/mono/mono/mini/ssa.c:365]:
```suggestion
		int stack_history_len = create_new_vars (cfg, max_vars, bb, originals, stack, lvreg_stack, lvreg_defined, &stack_history, &stack_history_size);
```

File: src/mono/mono/mini/ssa.c [hunk 4]
```diff
@@ -284,49 +307,103 @@ mono_ssa_rename_vars (MonoCompile *cfg, int max_vars, MonoBasicBlock *bb, gboole
...
-			else
-				/* The phi nodes are at the beginning of the bblock */
-				break;
-		}
-	}
 
-	if (bb->dominated) {
-		for (tmp = bb->dominated; tmp; tmp = tmp->next) {
-			mono_ssa_rename_vars (cfg, max_vars, (MonoBasicBlock *)tmp->data, originals_used, stack, lvreg_stack, lvreg_defined, stack_history + stack_history_len, stack_history_size - stack_history_len);
+			RenameStackInfo* info = (RenameStackInfo*)rename_stack + rename_stack_idx;
```
Comment (imhameed) [inline src/mono/mono/mini/ssa.c:374]:
```suggestion
			RenameStackInfo* info = rename_stack [rename_stack_idx];
```

File: src/mono/mono/mini/ssa.c [hunk 5]
```diff
@@ -284,49 +307,103 @@ mono_ssa_rename_vars (MonoCompile *cfg, int max_vars, MonoBasicBlock *bb, gboole
...
+			stack_history += stack_history_len;
+			stack_history_size -= stack_history_len;
+			blocks = bb->dominated;
+		} else {
+			restore_stack (stack, stack_history, stack_history_len);
+			blocks = blocks->next;
+
+			while (!blocks && rename_stack_idx > 0) {
+				RenameStackInfo* info = (RenameStackInfo*)rename_stack + rename_stack_idx - 1;
+				rename_stack_idx--;
```
Comment (imhameed) [inline src/mono/mono/mini/ssa.c:389]:
```suggestion
				rename_stack_idx--;
				RenameStackInfo* info = rename_stack [rename_stack_idx];
```

**Review by imhameed — CHANGES_REQUESTED** [review]
Looks fine. I've added some minor stylistic suggestions. And it would be nice if we had a dedicated dynamic array data structure (instead of needing to roll one from scratch, like is done here), but I don't think this PR should be blocked on that.

File: src/mono/mono/mini/ssa.c (hunk 4 above)
Comment (radekdoulik) [reply src/mono/mono/mini/ssa.c:373]: We need pointer here. I can get rid of the unneeded cast at least.

File: src/mono/mono/mini/ssa.c (hunk 5 above)
Comment (radekdoulik) [reply src/mono/mono/mini/ssa.c:388]: Same as above.

**Review by imhameed — APPROVED** [review]

**Pull request merged**